torchaudio/functional.py

```diff
@@ -1287,15 +1287,26 @@ def phaser(
     delay_pos = 0
     mod_pos = 0
 
-    output_waveform = torch.zeros_like(waveform, dtype=dtype, device=device)
+    # TODO: Autograd support?
+    output_waveform_pre_gain_list = []
+    waveform = waveform * gain_in
+    delay_buf = delay_buf * decay
+    waveform_list = [waveform[:, i] for i in range(waveform.size(1))]
+    delay_buf_list = [delay_buf[:, i] for i in range(delay_buf.size(1))]
+    mod_buf_list = [mod_buf[i] for i in range(mod_buf.size(0))]
```

Comment:
The gain seems to be from using lists instead of pytorch tensors, right?

Reply:
Yes

Reply:
Note however that the entries of this list are still torch.Tensors. The difference here is that they're created once ahead of time instead of repeatedly.

Also, much of the is from applying the scalar operations (such as gain or gain_out) onto the entire input or output tensors instead of per-loop, which is much, much faster.

Reply:
Hi @cpuhrsch , I understood the gain_in, gain_out optimizations , But is using the list of 1D tensor compared to 2D tensor giving additional performance gain ? When i did below experiment i am getting similar time for both list of 1D tensors and slicing 2D tensor
<img width="678" alt="Screenshot 2020-05-30 at 7 37 53 PM" src="https://user-images.githubusercontent.com/12258823/83330527-1dba1e00-a2ad-11ea-9c6a-a3f4f7fa6c07.png">





> The difference here is that they're created once ahead of time instead of repeatedly.

Both list of 1D tensor and 2D tensor represent same data before start of loop right . Am I missing something here ?
``` buf = torch.zeros(10,100000) and buf_list =  [buf[:,i] for i in range(100000)]```

Reply:
But also note that we're also converting delay_buf_list and mod_buf_list which would see a lot of repeated indexing and this way that won't be necessary. Indexing creates Tensor objects which is relatively expensive in comparison to the amount of compute being done.